### PR TITLE
Allow analyzer exceptions to fall through if desired

### PR DESCRIPTION
**Pull request by jjonescz**
Resolves https://github.com/dotnet/roslyn/issues/70660.

### CODE REVIEW

File: src/Compilers/Core/Portable/Compilation/Compilation.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: My main nit here is letting @jaredpar name things. I would prefer a flag on the lines of `disable-analyzer-exception-handling`. `debug-analyzer` is not particularly clear what it means.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs
Comment (333fred) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:1197]: Do we want to assert in the filter block, or just FailFast?

File: src/Compilers/Core/Portable/Compilation/Compilation.cs
Comment (jaredpar) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: The `debug-` prefix matches other feature flags used to debug compiler behavior. Like `debug-determinism`.

Comment (333fred) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: I suppose, but we're not debugging compiler behavior here. We're disabling compiler handling of exceptions from analyzers.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:1198]: >Environment.FailFast(CreateAnalyzerExceptionDiagnostic(analyzer, ex, info).ToString()); [](http://example.com/codeflow?start=16&length=87)

Is this really going to produce a Watson dump with the stack at the position that threw the exception and with the right exception information for it?

Comment (jaredpar) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:1198]: As long as this is called in the `when` then it will produce a Watson with failure on the stack.

File: src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs:68]: Why do we have a `!debugAnalyzers` here? My mental model is that when the flag is set that essetnially any `when` block where it's checked should be `FailFast` not rethrownig. Is there a subtly I'm missing with this one?

File: src/Compilers/Core/Portable/Compilation/Compilation.cs
Comment (jaredpar) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: It's not so much that we're disabling handling, we're aggressingly failing fast on them.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:1188]: Why do we need a second `HandleAnalyzer` function here? At a glance seems like the logic could just be rolled into the existing one. Doesn't seem like the other one is called anywhere else in the product (local search is acting up today though could be missing somtehnig)

File: src/Compilers/Core/Portable/Compilation/Compilation.cs
Comment (jaredpar) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: I do wonder if we should change the internal API name though. Seeing `DebugAnalyzers` in the code doesn't line up with the behavior as much as I would like.

Comment (333fred) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: I agree, and that's part of what makes me want to change the feature flag name as well: it doesn't convey what it actually does.

Comment (jaredpar) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: Feature flag names aren't meant to be user friendly. They're (mostly) undocumented switches we support, add and can delete at any time. I'd prefer they just follow a pattern.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs
Comment (jjonescz) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:1197]: Assert will present you with the nice "Attach debugger" dialog on .NET Framework which can be useful. Plus it might be friendlier that it throws an exception instead of failing fast in Debug config.

File: src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs
Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs:68]: We are not rethrowin. If `debugAnalyzers` is `true`, the `catch` won't be considered, so no exceptions are caught here, they instead propagate up and get handled by the `Environment.FailFast`. Could do the fail fast here instead, but not sure what would be the advantage of that?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs
Comment (jjonescz) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:1188]: The existing one is used in more places:

https://github.com/dotnet/roslyn/blob/5b73439b70c5c5263b5acbb69b426c0264f14af5/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs#L1214

https://github.com/dotnet/roslyn/blob/5b73439b70c5c5263b5acbb69b426c0264f14af5/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs#L1226

I think it's not worth propagating the feature flag to those places as they only evaluate properties `SupportedDiagnostics` and `SupportedSuppressions` (and looks like propagating the feature flag there would be complex as that code has access only to `CompilationOptions`, not `Compilation`).

File: src/Compilers/Core/Portable/Compilation/Compilation.cs
Comment (jjonescz) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: Seems like you want me to rename the property? What name do you suggest? I guess I could negate the semantics of the property and call it `CatchAnalyzerExceptions`.

Comment (333fred) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: > I'd prefer they just follow a pattern.

We don't have a pattern, we have _one_ example of something named `debug-`. AFAICT, here are all the `Feature` switches we have today:

* `strict`
* `disable-length-based-switch`
* `run-nullable-analysis`
* `peverify-compat`
* `debug-determinism`
* `pdb-path-determinism`
* `enable-generator-cache`
* `nullablePublicOnly`

> Feature flag names aren't meant to be user friendly.

Sure, but we're also users of the codebase. I'd prefer to have a flag that actually conveys what it does, rather than having to read through the code to determine what it actually does if I ever come across the flag and have forgotten what it means.

Comment (jaredpar) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: > I guess I could negate the semantics of the property and call it CatchAnalyzerExceptions.

That works for me. 

> We don't have a pattern, we have one example of something named debug-. 

How many of those are debugging compiler behavior though?

Comment (333fred) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: > How many of those are debugging compiler behavior though?

I still do not think that is a relevant question, as this new switch isn't debugging compiler behavior either 🙂.

Comment (jaredpar) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: We are ~8 comments deep on the name of a switch that will be used by a small number of customers when debugging edge cases in the compiler. We need to pick a name and get the functionality in. 

@jjonescz please make a decision.

File: src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs
Comment (jaredpar) [reply src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs:68]: My fear here is that it works the way you describe today. The two `catch` blocks feel a bit distant from each other. Worry that it would be easy for someone to inadvertently insert a `catch` in the middle that didn't respect the flag and broke the desired behavior. Do we have a test for this case hitting the `FailFast` code? If so I would be okay with the current behavior.

File: src/Compilers/Core/Portable/Compilation/Compilation.cs
Comment (jjonescz) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:3477]: Alright, I will change the property name, but leave the feature flag name (because I don't know how else to name it - `disable-analyzer-exception-handling` isn't accurate; `no-catch-analyzer-exceptions` seems like an antipattern with the `no-` prefix - plus I like short name for the flag).

**Review by 333fred — APPROVED** [review]

**Comment by jjonescz** [comment]
@chsienki @dotnet/roslyn-compiler for a second review, thanks

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs
Comment (chsienki) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs:318]: I'm curious why we bother creating the diagnostic at all in these cases? The diagnostic is just supposed to be a user-friendly representation of the exception, but here we actively _want_ the exception info. Do we actually gain anything by wrapping it in the diagnostic?

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs
Comment (chsienki) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs:16]: This feels incredible viral for such an edge-case change. Obviously, we need to thread it through somehow, but given this is a feature flag that is expected to be really a debugging only aid, could we instead just have a static field on `UserFunctionExtensions` that we just set in the ctor of the GeneratorDriver? 

I know its maybe not 'perfect' code, but it seems much less invasive for such a niche case

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs
Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs:318]: It has the advantage of including the source generator name (and maybe some other info could be added in the future, so feels like this could be worth keeping in sync, I think there is some context info for analyzers passed in for example).

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs
Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs:16]: I guess I would be fine with that, the question is: would I get a sign-off from @333fred on that, as well?

Comment (333fred) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs:16]: I feel like you never know when the next time we're going to want to plumb more options through the generator driver is going to be. If anything, what I would maybe do is pull the bool into a struct so that we can add more options later :).

I don't _hate_ the static state (we do that with `TestOnlyCompilationData`, for example), but I'm not a fan either.

Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs:16]: > I feel like you never know when the next time we're going to want to plumb more options through the generator driver is going to be.

👍 

> pull the bool into a struct so that we can add more options later

Reasonable, but I think that can be done when needed in the future.

**Comment by jjonescz** [comment]
@jaredpar @chsienki @dotnet/roslyn-compiler for a second review, thanks

**Review by jaredpar — APPROVED** [review]

**Pull request merged**
